### PR TITLE
gtk4-macros: Add blueprint support

### PR DESCRIPTION
**Pull request by YuraIz**

### CODE REVIEW

File: gtk4-macros/src/attribute_parser.rs [hunk 1]
```diff
@@ -15,7 +15,21 @@ pub struct Template {
...
-    String(String),
+    Xml(String),
+    Blueprint(String),
+}
+
+impl TemplateSource {
+    fn from_string_source(source: String) -> Result<Self> {
+        for c in source.chars() {
+            if c.is_ascii_alphabetic() {
+                return Ok(Self::Blueprint(source));
```
Comment (sdroege) [inline gtk4-macros/src/attribute_parser.rs:26]: That should probably be explicitly marked in the code which of the two it is

File: gtk4-macros/tests/templates.rs [hunk 2]
```diff
@@ -199,3 +199,57 @@ glib::wrapper! {
 fn template_child_without_attribute() {
     let _: MyWidget3 = glib::Object::new(&[]);
 }
+
+mod imp4 {
+    use super::*;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(string = "
+    template MyWidget4 : Widget {
```
Comment (sdroege) [inline gtk4-macros/tests/templates.rs:208]: blueprint is not installed on the CI so this would currently fail :) Can you add it?

File: gtk4-macros/src/blueprint.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+pub(crate) fn compile_blueprint(blueprint: &[u8]) -> Result<String> {
+    let mut compiler = Command::new("blueprint-compiler")
+        .args(&["compile", "-"])
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let mut stdin = compiler.stdin.take().unwrap();
+
+    stdin.write(b"using Gtk 4.0;\n")?;
```
[the remark this answers is not in the document]
Comment (YuraIz) [reply gtk4-macros/src/blueprint.rs:14]: I think it isn't needed to explicitly write gtk version because 4.0 is the only possible version, but user will be able to write "using Adw 1;" manually (blueprint doesn't warn about duplicate `using` lines)

[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply gtk4-macros/src/blueprint.rs:14]: Not sure this is needed at all... It is only useful for blueprint language server files to load the correct gir files and provide auto completion for you which is not something that is going to work in this case. Please drop it.

[the remark this answers is not in the document]
Comment (YuraIz) [reply gtk4-macros/src/blueprint.rs:14]: > Not sure this is needed at all... It is only useful for blueprint language server files to load the correct gir files and provide auto completion for you which is not something that is going to work in this case. Please drop it.

For some reason it fails when I don't use 'using' line
image

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -53,6 +53,13 @@ jobs:
         with:
           toolchain: ${{ matrix.rust }}
           override: true
+          
+      - run: |
+          # install blueprint-compiler
+          git clone --depth 1 https://gitlab.gnome.org/jwestman/blueprint-compiler
+          cd blueprint-compiler
+          meson _build
+          sudo ninja -C _build install
```
Comment (bilelmoussaoui) [inline .github/workflows/CI.yml:62]: Should be submitted as a separate PR and change the Dockerfile instead.

File: gtk4-macros/src/blueprint.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply gtk4-macros/src/blueprint.rs]: Sounds like a bug on blueprint side.

File: .github/workflows/CI.yml (hunk 4 above)
Comment (YuraIz) [reply .github/workflows/CI.yml]: I got that error with that CI file, but I can't get it locally

Screenshot from 2023-01-09 17-01-52

File: gtk4-macros/src/blueprint.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (YuraIz) [reply gtk4-macros/src/blueprint.rs]: Documentation says gtk declaration is required

image

File: .github/workflows/CI.yml (hunk 4 above)
Comment (YuraIz) [reply .github/workflows/CI.yml]: #1239

**Comment by YuraIz** [comment]
@bilelmoussaoui I don't understand why CI fails,

CI on that PR can't find blueprint-compiler, but action at the same commit from my repo fails with a diffrent error

https://github.com/YuraIz/gtk4-rs/actions/runs/3875198805/jobs/6607918824

**Comment by bilelmoussaoui** [comment]
How are you running the tests locally?

**Comment by YuraIz** [comment]
That's command copied from CI
image

It works the same with `cargo t` in the gtk4-macros folder
image

**Comment by bilelmoussaoui** [comment]
you have to enable the xml_validation feature to see the breakage.

**Comment by YuraIz** [comment]
image
image

**Comment by YuraIz** [comment]
I don't want to add blueprint-compiler to windows test until it will be awailable as pip package

**Comment by bilelmoussaoui** [comment]
Thanks!

**Comment by vanillajonathan** [comment]
Does this work with template resources loaded from files too?

**Comment by bilelmoussaoui** [comment]
No that is not possible.
